server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/SMTPBehaviorRepository.java

```diff
@@ -19,36 +19,98 @@
+                .getNumberOfAnswers()
+                .map(numAnswers -> new MockSMTPBehaviorInformation(behavior, Optional.of(numAnswers)))
+                .orElseGet(() -> new MockSMTPBehaviorInformation(behavior, Optional.empty()));
+        }
+
+        private final MockSMTPBehavior behavior;
+        private Optional<Integer> remainingAnswers;
+
+        MockSMTPBehaviorInformation(MockSMTPBehavior behavior, Optional<Integer> remainingAnswers) {
+            this.behavior = behavior;
+            this.remainingAnswers = remainingAnswers;
+        }
+
+        void decreaseRemainingAnswers() {
+            remainingAnswers = remainingAnswers
+                .map(remainingAnswer -> remainingAnswer - 1);
+        }
+
+        MockSMTPBehavior getBehavior() {
+            return behavior;
+        }
+
+        Optional<Integer> getRemainingAnswers() {
+            return remainingAnswers;
+        }
+    }
+
+    private final ConcurrentLinkedQueue<MockSMTPBehaviorInformation> behaviorsInformation;
+
+    SMTPBehaviorRepository() {
+        this.behaviorsInformation = new ConcurrentLinkedQueue<>();
+    }
+
+    void clearBehaviors() {
+        this.behaviorsInformation.clear();
+    }
+
+    void setBehaviors(MockSmtpBehaviors behaviors) {
+        behaviors.getBehaviorList()
+            .stream()
+            .map(MockSMTPBehaviorInformation::from)
+            .forEach(behaviorsInformation::add);
+    }
 
-    public SMTPBehaviorRepository() {
-        this.behaviors = new AtomicReference<>();
+    Stream<MockSMTPBehaviorInformation> behaviorsInformation() {
+        return behaviorsInformation.stream();
     }
 
-    public Optional<MockSmtpBehaviors> getBehaviors() {
-        return Optional.ofNullable(this.behaviors.get());
+    Stream<MockSMTPBehavior> allBehaviors() {
+        return behaviorsInformation()
+            .map(MockSMTPBehaviorInformation::getBehavior);
     }
 
-    public void clearBehaviors() {
-        this.behaviors.set(null);
+    Optional<MockSMTPBehaviorInformation> behaviorInformation(MockSMTPBehavior behavior) {
+        return behaviorsInformation()
+            .filter(behaviorInformation -> behaviorInformation.getBehavior().equals(behavior))
+            .findFirst();
     }
 
-    public void setBehaviors(MockSmtpBehaviors behaviors) {
-        this.behaviors.set(behaviors);
+    void decreaseRemainingAnswers(MockSMTPBehavior behavior) {
+        behaviorsInformation()
+            .filter(behaviorInformation -> behaviorInformation.getBehavior().equals(behavior))
+            .findFirst()
+            .ifPresent(this::decreaseRemainingAnswers);
+    }
+
+    private void decreaseRemainingAnswers(MockSMTPBehaviorInformation behaviorInformation) {
+        behaviorInformation.decreaseRemainingAnswers();
+
+        behaviorInformation.remainingAnswers
+            .filter(numAnswers -> numAnswers == NO_ANSWER_LEFT)
+            .ifPresent(any -> behaviorsInformation.remove(behaviorInformation));
```

Comment:
I'm against that side effect.

The behavior should still be here, but with 0 remaining answers.